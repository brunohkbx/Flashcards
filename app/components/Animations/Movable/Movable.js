import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Animated } from 'react-native';
import { withTheme } from 'react-native-paper';

export class Movable extends Component {
  state = {
    translateY: new Animated.Value(0)
  };

  animatedStyle = () => {
    const {
      toValue,
      theme: { transitions: { duration, easing }}
    } = this.props;

    return {
      toValue,
      duration: toValue < 0 ? duration.enteringScreen : duration.leavingScreen,
      easing: toValue < 0 ? easing.easeOut : easing.sharp,
      useNativeDriver: true
    };
  };

  animate = () => {
    const { translateY } = this.state;

    Animated.timing(translateY, this.animatedStyle()).start();
  }

  componentDidMount = () => {
    this.animate();
  }

  componentDidUpdate(prevProps) {
    if (this.props.toValue !== prevProps.toValue)
      this.animate();
  }

  render() {
    const { children, style } = this.props;
    const { translateY } = this.state;

    return (
      <Animated.View style={{ transform: [{ translateY }] }} >
        {children}
      </Animated.View>
    );
  }
}

Movable.propTypes = {
  toValue: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired
};

export default withTheme(Movable);
