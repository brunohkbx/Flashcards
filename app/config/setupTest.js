import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
require('stacktrace-parser');

// React 16 Enzyme adapter
configure({ adapter: new Adapter() });

jest.mock('NativeAnimatedHelper');
import 'react-native/Libraries/Animated/src/bezier';
