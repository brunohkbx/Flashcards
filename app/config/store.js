import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'remote-redux-devtools';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';

const composeEnhancers = composeWithDevTools({ realtime: true, port: 8000 });

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk))
);

export default store;
