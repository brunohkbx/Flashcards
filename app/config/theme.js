import { Easing } from 'react-native';
import { DefaultTheme, Colors } from 'react-native-paper';

export const primaryColor = DefaultTheme.colors.primary;
export const white = Colors.white;

export const theme = {
  ...DefaultTheme,
  transitions: {
    duration: {
      enteringScreen: 225,
      leavingScreen: 195
    },
    easing: {
      easeOut: Easing.bezier(0.0, 0, 0.2, 1),
      sharp: Easing.bezier(0.4, 0, 0.6, 1)
    },
  }
};
