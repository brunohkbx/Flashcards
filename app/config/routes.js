import { createStackNavigator } from 'react-navigation';
import Main from '../screens/Main';
import CreateDeck from '../screens/CreateDeck';
import EditDeck from '../screens/EditDeck';
import DeckDetail from '../screens/DeckDetail';
import Quiz from '../screens/Quiz/Quiz';
import { primaryColor, white } from './theme';

export const RootStack =  createStackNavigator(
  {
    Main: {
      screen: Main,
      navigationOptions: {
        title: 'Flashcards'
      }
    },
    CreateDeck: {
      screen: CreateDeck,
      navigationOptions: {
        title: 'Create Deck'
      }
    },
    EditDeck: {
      screen: EditDeck,
      navigationOptions: {
        title: 'Edit Deck'
      }
    },
    DeckDetail: {
      screen: DeckDetail
    },
    Quiz: {
      screen: Quiz,
      navigationOptions: {
        title: 'Quiz'
      }
    }
  },
  {
    initialRouteName: 'Main',
    navigationOptions: {
      headerStyle: {
        backgroundColor: primaryColor,
      },
      headerTintColor: white
    }
  }
);
