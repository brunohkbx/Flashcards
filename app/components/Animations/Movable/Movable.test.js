import React from 'react';
import { shallow } from 'enzyme';
import { Movable } from './Movable';
import { Animated, View } from 'react-native';
import { theme } from '../../../config/theme';

describe('Movable', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      toValue: 0,
      theme
    }, propOverrides);

    const wrapper = shallow(<Movable {...defaultProps} ><View /></Movable>);
    const wrapperInstance = wrapper.instance();

    return {
      wrapper, wrapperInstance
    };
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe('animatedStyle', () => {
    it('performs a enteringScreen animation when toValue is less than 0', () => {
      const { wrapperInstance } = setup({ toValue: -46 });
      const { transitions: { duration, easing }} = theme;

      expect(wrapperInstance.animatedStyle()).toEqual(
        {
          toValue: -46,
          duration: duration.enteringScreen,
          easing: easing.easeOut,
          useNativeDriver: true
        }
      );
    });

    it('performs a leavingScreen animation when toValue is greater or equal to 0', () => {
      const { wrapperInstance } = setup({ toValue: 0 });
      const { transitions: { duration, easing }} = theme;

      expect(wrapperInstance.animatedStyle()).toEqual(
        {
          toValue: 0,
          duration: duration.leavingScreen,
          easing: easing.sharp,
          useNativeDriver: true
        }
      );
    });
  });

  describe('animate', () => {
    it('starts the animation', () => {
      const { wrapperInstance } = setup();
      jest.spyOn(Animated, 'timing');

      wrapperInstance.animate();

      expect(Animated.timing).toHaveBeenCalledWith(
        wrapperInstance.state.translateY,
        wrapperInstance.animatedStyle()
      );
    });
  });

  describe('componentDidUpdate', () => {
    it('calls animate again if there is a new value for toValue', () => {
      const { wrapper, wrapperInstance } = setup();
      jest.spyOn(wrapperInstance, 'animate');

      wrapper.setProps({ toValue: 50 });

      expect(wrapperInstance.animate).toHaveBeenCalled();
    });

    it('does not call animate if there is not a new value for toValue', () => {
      const { wrapper, wrapperInstance } = setup();
      jest.spyOn(wrapperInstance, 'animate');

      wrapper.setProps({ toValue: 0 });

      expect(wrapperInstance.animate).not.toHaveBeenCalled();
    });
  });
});
