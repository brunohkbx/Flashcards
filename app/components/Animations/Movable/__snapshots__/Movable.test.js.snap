// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movable renders properly 1`] = `
<AnimatedComponent
  style={
    Object {
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
    }
  }
>
  <Component />
</AnimatedComponent>
`;
